src/opm/parser/eclipse/EclipseState/Grid/MapAxes.cpp

```diff
@@ -0,0 +1,156 @@
+/*
+  Copyright 2021 Equinor ASA.
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <cmath>
+#include <string>
+
+#include <opm/common/utility/String.hpp>
+#include <opm/parser/eclipse/Units/Units.hpp>
+#include <opm/parser/eclipse/EclipseState/Grid/MapAxes.hpp>
+#include <opm/parser/eclipse/Deck/Deck.hpp>
+#include <opm/io/eclipse/EclFile.hpp>
+#include <opm/io/eclipse/EclOutput.hpp>
+#include <opm/parser/eclipse/Parser/ParserKeywords/M.hpp>
+
+namespace Opm {
+
+namespace {
+
+double length_factor(const std::string& mapunits) {
+    auto mp = trim_copy(mapunits);
+    if (mp == "METRES")
+        return unit::meter;
+    else if (mp == "FEET")
+        return unit::feet;
+    else if (mp == "CM")
+        return prefix::centi*unit::meter;
+    throw std::logic_error("Unrecognized MAPUNITS value: " + mapunits);
+}
+
+}
+
+
+
+MapAxes::MapAxes() :
+    origo({0,0}),
+    unit_x({1,0}),
+    unit_y({0,1}),
+    m_input({1,0,0,0,0,1}),
+    inv_norm(1.0)
+{
+}
+
+
+const std::optional<std::string>& MapAxes::mapunits() const {
+    return this->map_units;
+}
+
+
+const std::vector<float>& MapAxes::input() const {
+    return this->m_input;
+}
+
+
+void MapAxes::init(double length_factor, double X1, double Y1, double X2, double Y2, double X3, double Y3)
+{
+    auto f = [](double d) { return static_cast<float>(d); };
+    this->m_input = {f(X1),f(Y1),f(X2),f(Y2),f(X3),f(Y3)};
+    this->origo = { length_factor*X2, length_factor*Y2};
+    this->unit_x = {X3 - X2, Y3 - Y2};
+    this->unit_y = {X1 - X2, Y1 - Y2};
+
+    auto norm_x = 1.0/std::sqrt( this->unit_x[0]*this->unit_x[0] + this->unit_x[1]*this->unit_x[1] );
+    auto norm_y = 1.0/std::sqrt( this->unit_y[0]*this->unit_y[0] + this->unit_y[1]*this->unit_y[1] );
```

Comment:
In C++17 you can use [`std::hypot()`](https://en.cppreference.com/w/cpp/numeric/math/hypot) to compute these denominators.  That will usually be more accurate.
```C++
const auto norm_x = 1.0 / std::hypot(this->unit_x[0], this->unit_x[1]);
const auto norm_y = 1.0 / std::hypot(this->unit_y[0], this->unit_y[1]);
```

Reply:
> In C++17 you can use std::hypot() 

Had not heard about that one - std::c++ sure has a function for "everything".

> That will usually be more accurate.

Only in case the standard version underflows/overflows? I'll update anyway.

Reply:
> In C++17...

That was a misleading statement.  C++11 introduced `std::hypot()`.  C++17 introduced the _three parameter_ overload of the function.